rust/src/ops/max/mod.rs

```diff
@@ -0,0 +1,111 @@
+use std::cmp::{Ord, PartialEq};
+use std::marker::PhantomData;
+
+use num_traits::cast::{ToPrimitive, NumCast};
+
+use alga::general::AbstractMagma;
+use alga::general::AbstractMonoid;
+use alga::general::AbstractSemigroup;
+use alga::general::Identity;
+use alga::general::Operator;
+
+use super::AggregateOperator;
+use super::AggregateMonoid;
+
+// We need a generic way to talk about a type's minimum 
+// value. Unfortunately, we'll need a new implementation 
+// for every type we want to support. I don't know of a 
+// way around this yet.
+pub trait Min {
+    fn min() -> Self;
+}
+
+impl Min for i32 {
+    fn min() -> i32 {
+        i32::MIN
+    }
+}
+
+/// Binary operator for maximum.
+/// Has the following properties:
+/// * Associativity
+/// * Commutativity
+#[derive(Copy, Clone)]
+pub struct Max<In, Out> {
+    in_type: PhantomData<In>,
+    out_type: PhantomData<Out>
+}
+
+// We need this because of the orphan rules for trait implemenations. See 
+// `rustc --explain E0210`.
+#[derive(Copy, Clone, Eq, PartialEq)]
+pub struct MaxPartial<T> {
+    val: T
+}
```

Comment:
For later, there is a crate for implementing new-types which I use frequently called [`shrinkwraprs`](https://docs.rs/shrinkwraprs/0.3.0/shrinkwraprs/). I hate typing the name because I always misspell it, but the crate is pretty useful. With it you could do:

```rust
#[derive(shrinkwraprs::Shrinkwrap, Copy, Clone, Eq, PartialEq)]
pub struct MaxPartial<T>(pub T);
```
and you would be able to use `T` without `.val`.

Reply:
Excellent, very promising. I wanted to go with a tuple initially, but I thought `.val` was less strange than `.0`.